proposed/0008-Connection_management.md

```diff
@@ -0,0 +1,82 @@
+- Feature Name: Improved Connection Management
+- Type: Enhancement
+- Related Components: routing, crust
+- Start Date: 23-09-2015
+- RFC PR: (leave this empty)
+- Issue Number: (leave this empty)
+
+# Summary
+
+Introduce the notion of clearly defined connection phases and establish directed routing network connections in conjunction with crust.
+
+# Motivation
+
+Both outgoing and incoming connections are created to peers with unknown direction. Use the connect/accept connection code available from crust to store timed state in-line with crust connection handling changes for connections in either direction. This will allow us to determine unambiguously who the connection initiator is and act accordingly.
+
+# Detailed design
+
+## State
+
+Introduce a `State` object to `RoutingCore` representing the distinct phases of execution for network entities characterised as follows.
+
+1. Initially client/node in disconnected state, and when all connections are dropped/lost.
+1. Bootstrapping, initiated in constructor by calling crust service function and halted by crust event sent over channel.
+1. If non-client node, connected phase adds connections to routing table.
+1. In order to prevent any further network activity a terminated state.
+
+```rust
+pub enum Phase {
+    Disconnected(bool),
+    Bootstrapping(bool),
+    Connected(bool),
+    Terminated(bool),
+}
+```
```

Comment:
Doesn't `Bootstrapping` imply `Disconnected`?  If so, can we remove `Bootstrapping`?

Apart from that, these states are mutually exclusive aren't they?  If so, why do we need the `(bool)` aspect to each - wouldn't a C-style enum do?